src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,792 @@
+                            SyntaxNode newParentExpressionStatement = parentStatement.ReplaceNode(source, localFunctionInvocation.WithAdditionalAnnotations(Simplifier.Annotation));
+                            documentUpdateInfo = new DocumentUpdateInfo(parentStatement, new[] { localFunctionDeclaration, newParentExpressionStatement });
+                            return true;
+                        }
+                    }
+                }
+
+                documentUpdateInfo = default;
+                return false;
+            }
+
+            private bool TryGetTypeSyntax(ExpressionSyntax expression, out TypeSyntax typeSyntax)
+            {
+                var typeSymbol = _semanticModel.GetTypeInfo(expression, _cancellationToken).Type;
+                if (typeSymbol.TypeKind == TypeKind.Error ||
+                    typeSymbol.ContainsAnonymousType())
+                {
+                    typeSyntax = default;
+                    return false;
+                }
+
+                typeSyntax = typeSymbol.GenerateTypeSyntax();
+                return true;
+            }
+
+            private string GetFreeSymbolName(string prefix, int position)
+                => NameGenerator.GenerateUniqueName(
+                    prefix,
+                    _semanticModel.LookupSymbols(position).Select(symbol => symbol.Name).ToImmutableHashSet(),
+                    StringComparer.CurrentCulture);
+
+            private bool TryConvertIfInForEach(QueryExpressionSyntax source, ForEachStatementSyntax forEachStatement, out DocumentUpdateInfo documentUpdateInfo)
+            {
+                // before foreach(var x in from a in b select a)
+                var statement = forEachStatement.Statement;
+                StatementSyntax internalNodeMethod(ExpressionSyntax expression)
+                {
+                    // before 
+                    //  foreach (var a in from a in b where a > 5 select a) 
+                    //  { 
+                    //      dosomething(a); 
+                    //  }
+                    // after 
+                    //  foreach (var a in b)
+                    //  {
+                    //      if (a > 5)
+                    //      {
+                    //          dosomething(a); 
+                    //      }
+                    //  }
+                    if (expression is IdentifierNameSyntax identifier &&
+                        identifier.Identifier.ValueText == forEachStatement.Identifier.ValueText)
+                    {
+                        return statement.WithAdditionalAnnotations(Formatter.Annotation);
+                    }
+                    else
+                    {
+                        // before foreach(var x in from ... a) { dosomething(x); }
+                        // after 
+                        // foreach (var a in ...)
```

Comment:
>var a [](start = 36, length = 5)

Do we have a tests when query produces a sequence of Anonymous Types?

Reply:
yes, AnonymousType. Please let me know if we need more.

---
In reply to: [177202031](https://github.com/dotnet/roslyn/pull/25362#discussion_r177202031) [](ancestors = 177202031)